File: apps/android/app/src/main/java/org/llamenos/hotline/service/TranscriptionService.kt
package org.llamenos.hotline.service

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.speech.RecognitionListener
import android.speech.RecognizerIntent
import android.speech.SpeechRecognizer
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * State representing the current transcription status.
 */
sealed interface TranscriptionState {
    /** Service idle, not transcribing. */
    data object Idle : TranscriptionState

    /** Actively transcribing audio. */
    data object Transcribing : TranscriptionState

    /** Transcription stopped, final result available. */
    data object Stopped : TranscriptionState

    /** An error occurred during transcription. */
    data class Error(val message: String) : TranscriptionState
}

/**
 * On-device speech transcription using Android's SpeechRecognizer.
 * Audio never leaves the device — all processing uses on-device recognition
 * when available (Android 12+). Supports real-time streaming transcription
 * during calls.
 *
 * Injected as a singleton via Hilt. Manages SpeechRecognizer lifecycle
 * and emits transcription results via StateFlows.
 */
@Singleton
class TranscriptionService @Inject constructor(
    @ApplicationContext private val context: Context,
) {
    // MARK: - Public State

    private val _state = MutableStateFlow<TranscriptionState>(TranscriptionState.Idle)
    val state: StateFlow<TranscriptionState> = _state.asStateFlow()

    private val _liveTranscript = MutableStateFlow("")
    val liveTranscript: StateFlow<String> = _liveTranscript.asStateFlow()

    private val _finalTranscript = MutableStateFlow("")
    val finalTranscript: StateFlow<String> = _finalTranscript.asStateFlow()

    /** Whether speech recognition is available on this device. */
    val isAvailable: Boolean
        get() = SpeechRecognizer.isRecognitionAvailable(context)

    /** Whether on-device recognition is available (Android 12+). */
    val isOnDeviceAvailable: Boolean
        get() = android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S &&
            SpeechRecognizer.isOnDeviceRecognitionAvailable(context)

    // MARK: - Preferences

    /** Whether transcription is enabled by the user. */
    var isEnabled: Boolean
        get() = prefs.getBoolean(PREF_ENABLED, false)
        set(value) = prefs.edit().putBoolean(PREF_ENABLED, value).apply()

    /** Selected language for transcription. Empty string = auto-detect. */
    var selectedLanguage: String
        get() = prefs.getString(PREF_LANGUAGE, "") ?: ""
        set(value) = prefs.edit().putString(PREF_LANGUAGE, value).apply()

    // MARK: - Private

    private var speechRecognizer: SpeechRecognizer? = null
    private val accumulatedTranscript = StringBuilder()
    private val prefs by lazy {
        context.getSharedPreferences("transcription_prefs", Context.MODE_PRIVATE)
    }

    /**
     * Start real-time on-device transcription.
     * Audio is captured from the microphone and processed locally via SpeechRecognizer.
     * The [liveTranscript] flow updates in real-time as speech is recognized.
     *
     * Caller must have already obtained RECORD_AUDIO permission.
     */
    fun startTranscription() {
        if (_state.value == TranscriptionState.Transcribing) return

        if (!isAvailable) {
            _state.value = TranscriptionState.Error(
                context.getString(org.llamenos.hotline.R.string.transcription_speech_recognition_unavailable)
            )
            return
        }

        accumulatedTranscript.clear()
        _liveTranscript.value = ""
        _finalTranscript.value = ""

        val recognizer = if (isOnDeviceAvailable) {
            SpeechRecognizer.createOnDeviceSpeechRecognizer(context)
        } else {
            SpeechRecognizer.createSpeechRecognizer(context)
        }

        recognizer.setRecognitionListener(object : RecognitionListener {
            override fun onReadyForSpeech(params: Bundle?) {
                _state.value = TranscriptionState.Transcribing
            }

            override fun onBeginningOfSpeech() {}

            override fun onRmsChanged(rmsdB: Float) {}

            override fun onBufferReceived(buffer: ByteArray?) {}

            override fun onEndOfSpeech() {
                // SpeechRecognizer auto-stops after silence.
                // Restart if we're still in transcribing state to keep capturing.
                if (_state.value == TranscriptionState.Transcribing) {
                    restartListening(recognizer)
                }
            }

            override fun onError(error: Int) {
                when (error) {
                    SpeechRecognizer.ERROR_NO_MATCH,
                    SpeechRecognizer.ERROR_SPEECH_TIMEOUT -> {
                        // No speech detected — restart listening if still active
                        if (_state.value == TranscriptionState.Transcribing) {
                            restartListening(recognizer)
                        }
                    }
                    SpeechRecognizer.ERROR_CLIENT -> {
                        // Client-side error, usually from cancellation — ignore if stopping
                        if (_state.value == TranscriptionState.Transcribing) {
                            _state.value = TranscriptionState.Error(
                                context.getString(org.llamenos.hotline.R.string.transcription_transcription_error)
                            )
                        }
                    }
                    else -> {
                        _state.value = TranscriptionState.Error(
                            context.getString(org.llamenos.hotline.R.string.transcription_transcription_error)
                        )
                    }
                }
            }

            override fun onResults(results: Bundle?) {
                val matches = results?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
                val text = matches?.firstOrNull() ?: return

                if (text.isNotBlank()) {
                    if (accumulatedTranscript.isNotEmpty()) {
                        accumulatedTranscript.append(" ")
                    }
                    accumulatedTranscript.append(text)
                    _liveTranscript.value = accumulatedTranscript.toString()
                }

                // Restart listening to continue capturing
                if (_state.value == TranscriptionState.Transcribing) {
                    restartListening(recognizer)
                }
            }

            override fun onPartialResults(partialResults: Bundle?) {
                val matches = partialResults?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
                val text = matches?.firstOrNull() ?: return

                // Show accumulated + current partial
                val current = if (accumulatedTranscript.isNotEmpty()) {
                    "${accumulatedTranscript} $text"
                } else {
                    text
                }
                _liveTranscript.value = current
            }

            override fun onEvent(eventType: Int, params: Bundle?) {}
        })

        speechRecognizer = recognizer
        startListening(recognizer)
    }

    /**
     * Stop transcription and finalize the transcript.
     * Returns the accumulated transcript text.
     */
    fun stopTranscription(): String {
        speechRecognizer?.stopListening()
        speechRecognizer?.cancel()
        speechRecognizer?.destroy()
        speechRecognizer = null

        val transcript = _liveTranscript.value
        _finalTranscript.value = transcript
        _state.value = TranscriptionState.Stopped

        return transcript
    }

    /** Reset the service to idle, clearing all transcript data. */
    fun reset() {
        stopTranscription()
        _liveTranscript.value = ""
        _finalTranscript.value = ""
        accumulatedTranscript.clear()
        _state.value = TranscriptionState.Idle
    }

    // MARK: - Private Helpers

    private fun createRecognizerIntent(): Intent {
        return Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
            putExtra(
                RecognizerIntent.EXTRA_LANGUAGE_MODEL,
                RecognizerIntent.LANGUAGE_MODEL_FREE_FORM,
            )
            putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, true)
            putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 1)

            // Set language preference
            val language = selectedLanguage
            if (language.isNotEmpty()) {
                putExtra(RecognizerIntent.EXTRA_LANGUAGE, language)
                putExtra(RecognizerIntent.EXTRA_LANGUAGE_PREFERENCE, language)
            }

            // Request on-device recognition for privacy (Android 13+)
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU) {
                putExtra(RecognizerIntent.EXTRA_PREFER_OFFLINE, true)
            }
        }
    }

    private fun startListening(recognizer: SpeechRecognizer) {
        try {
            recognizer.startListening(createRecognizerIntent())
        } catch (e: Exception) {
            _state.value = TranscriptionState.Error(e.message ?: "Failed to start listening")
        }
    }

    private fun restartListening(recognizer: SpeechRecognizer) {
        try {
            recognizer.startListening(createRecognizerIntent())
        } catch (_: Exception) {
            // If restart fails, that's okay — user can restart manually
        }
    }

    companion object {
        private const val PREF_ENABLED = "transcription_enabled"
        private const val PREF_LANGUAGE = "transcription_language"
    }
}

File: apps/ios/Sources/Services/TranscriptionService.swift
import Foundation
import Speech
import AVFoundation

// MARK: - TranscriptionState

/// Observable state for transcription progress and results.
enum TranscriptionState: Equatable {
    /// Service idle, not transcribing.
    case idle
    /// Requesting permissions.
    case requestingPermission
    /// Actively transcribing audio.
    case transcribing
    /// Transcription stopped, final result available.
    case stopped
    /// An error occurred.
    case error(String)
}

// MARK: - TranscriptionService

/// On-device speech transcription using Apple's Speech framework.
/// Audio never leaves the device — all processing is local (on-device mode).
/// Supports real-time streaming transcription during calls.
@Observable
final class TranscriptionService {
    // MARK: - Public State

    /// Current transcription state.
    var state: TranscriptionState = .idle

    /// Live transcript text, updated in real-time during transcription.
    var liveTranscript: String = ""

    /// Final transcript after transcription stops.
    var finalTranscript: String = ""

    /// Whether transcription is enabled in user preferences.
    var isEnabled: Bool {
        get { UserDefaults.standard.bool(forKey: "transcription_enabled") }
        set { UserDefaults.standard.set(newValue, forKey: "transcription_enabled") }
    }

    /// Selected transcription language locale identifier (e.g. "en-US", "es-ES").
    /// Empty string means auto-detect.
    var selectedLanguage: String {
        get { UserDefaults.standard.string(forKey: "transcription_language") ?? "" }
        set { UserDefaults.standard.set(newValue, forKey: "transcription_language") }
    }

    /// Whether speech recognition is available on this device.
    var isAvailable: Bool {
        SFSpeechRecognizer()?.isAvailable ?? false
    }

    /// Supported locale identifiers for speech recognition.
    var supportedLocales: [Locale] {
        Array(SFSpeechRecognizer.supportedLocales()).sorted { $0.identifier < $1.identifier }
    }

    // MARK: - Private

    private var speechRecognizer: SFSpeechRecognizer?
    private var recognitionRequest: SFSpeechAudioBufferRecognitionRequest?
    private var recognitionTask: SFSpeechRecognitionTask?
    private var audioEngine: AVAudioEngine?

    // MARK: - Permission

    /// Check and request speech recognition + microphone permissions.
    /// Returns true if both are granted.
    func requestPermissions() async -> Bool {
        state = .requestingPermission

        // Request speech recognition authorization
        let speechAuthorized = await withCheckedContinuation { continuation in
            SFSpeechRecognizer.requestAuthorization { status in
                continuation.resume(returning: status == .authorized)
            }
        }

        guard speechAuthorized else {
            state = .error(NSLocalizedString("transcription_mic_permission_denied", comment: ""))
            return false
        }

        // Request microphone permission
        let micAuthorized: Bool
        if #available(iOS 17, *) {
            micAuthorized = await AVAudioApplication.requestRecordPermission()
        } else {
            micAuthorized = await withCheckedContinuation { continuation in
                AVAudioSession.sharedInstance().requestRecordPermission { granted in
                    continuation.resume(returning: granted)
                }
            }
        }

        guard micAuthorized else {
            state = .error(NSLocalizedString("transcription_mic_permission_denied", comment: ""))
            return false
        }

        state = .idle
        return true
    }

    // MARK: - Start Transcription

    /// Start real-time on-device transcription.
    /// Audio is captured from the microphone and processed locally via Apple Speech framework.
    /// The `liveTranscript` property updates in real-time as speech is recognized.
    func startTranscription() async throws {
        guard state != .transcribing else { return }

        // Verify permissions
        let hasPermission = await requestPermissions()
        guard hasPermission else { return }

        // Configure speech recognizer
        let locale: Locale
        if selectedLanguage.isEmpty {
            locale = Locale.current
        } else {
            locale = Locale(identifier: selectedLanguage)
        }

        guard let recognizer = SFSpeechRecognizer(locale: locale), recognizer.isAvailable else {
            state = .error(NSLocalizedString("transcription_speech_recognition_unavailable", comment: ""))
            return
        }

        // Require on-device recognition for privacy
        if #available(iOS 13, *) {
            recognizer.supportsOnDeviceRecognition = true
        }

        speechRecognizer = recognizer
        liveTranscript = ""
        finalTranscript = ""

        // Create recognition request configured for on-device only
        let request = SFSpeechAudioBufferRecognitionRequest()
        request.shouldReportPartialResults = true

        if #available(iOS 13, *) {
            request.requiresOnDeviceRecognition = true
        }

        recognitionRequest = request

        // Configure audio session
        let audioSession = AVAudioSession.sharedInstance()
        try audioSession.setCategory(.record, mode: .measurement, options: .duckOthers)
        try audioSession.setActive(true, options: .notifyOthersOnDeactivation)

        // Set up audio engine
        let engine = AVAudioEngine()
        let inputNode = engine.inputNode
        let recordingFormat = inputNode.outputFormat(forBus: 0)

        inputNode.installTap(onBus: 0, bufferSize: 1024, format: recordingFormat) { [weak self] buffer, _ in
            self?.recognitionRequest?.append(buffer)
        }

        engine.prepare()
        try engine.start()
        audioEngine = engine

        // Start recognition task
        recognitionTask = recognizer.recognitionTask(with: request) { [weak self] result, error in
            guard let self else { return }

            if let result {
                Task { @MainActor in
                    self.liveTranscript = result.bestTranscription.formattedString

                    if result.isFinal {
                        self.finalTranscript = result.bestTranscription.formattedString
                        self.stopTranscriptionInternal()
                    }
                }
            }

            if let error {
                // Ignore cancellation errors (expected when stopping)
                let nsError = error as NSError
                if nsError.domain == "kAFAssistantErrorDomain" && nsError.code == 216 {
                    // "kAFAssistantErrorDomain error 216" = request was canceled
                    return
                }

                Task { @MainActor in
                    self.state = .error(error.localizedDescription)
                    self.stopTranscriptionInternal()
                }
            }
        }

        state = .transcribing
    }

    // MARK: - Stop Transcription

    /// Stop transcription and finalize the transcript.
    /// Returns the final transcript text.
    @discardableResult
    func stopTranscription() -> String {
        stopTranscriptionInternal()

        if finalTranscript.isEmpty {
            finalTranscript = liveTranscript
        }

        return finalTranscript
    }

    /// Reset the service to idle state, clearing all transcript data.
    func reset() {
        stopTranscriptionInternal()
        liveTranscript = ""
        finalTranscript = ""
        state = .idle
    }

    // MARK: - Private Helpers

    private func stopTranscriptionInternal() {
        audioEngine?.stop()
        audioEngine?.inputNode.removeTap(onBus: 0)
        audioEngine = nil

        recognitionRequest?.endAudio()
        recognitionRequest = nil

        recognitionTask?.cancel()
        recognitionTask = nil

        if state == .transcribing {
            state = .stopped
        }

        // Deactivate audio session
        try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
    }
}
